webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/NodeConfigurationService.scala

```diff
@@ -0,0 +1,350 @@
+   */
+  private[this] def performAction(action: CacheComplianceQueueAction): IOResult[Unit] = {
+    import CacheComplianceQueueAction._
+    // in a semaphore
+    semaphore.withPermit(
+       action match {
+                          case insert: InsertNodeInCache => IOResult.effectNonBlocking {  cache = cache + (insert.nodeId -> None) }
+                          case delete: RemoveNodeInCache => IOResult.effectNonBlocking {  cache = cache.removed(delete.nodeId) }
+                          case update: UpdateNodeConfiguration => IOResult.effectNonBlocking {  cache = cache + (update.nodeId -> Some(update.nodeConfiguration)) }
+                          case something =>
+                            Inconsistency(s"NodeConfiguration service cache received unknown command : ${something}").fail
+                           // should not happen
+                   }
+    )
+  }
+
+  /**
+   * invalidate with an action to do something
+   * order is important
+   */
+  def invalidateWithAction(actions: Seq[(NodeId, CacheComplianceQueueAction)]): IOResult[Unit] = {
+    ZIO.when(actions.nonEmpty) {
+      ReportLoggerPure.Cache.debug(s"Node Configuration cache: invalidation request for nodes with action: [${actions.map(_._1).map { _.value }.mkString(",")}]") *>
+        invalidateMergeUpdateSemaphore.withPermit(for {
+          elements     <- invalidateNodeConfigurationRequest.takeAll
+          allActions   =  (elements.flatten ++ actions)
+          _            <- invalidateNodeConfigurationRequest.offer(allActions)
+        } yield ())
+    }
+  }
+  // ? question ?
+  // how to properly ensure that cache is synchro ?
+  // we have the begin date of the nodeexpectedreport that my offer a way to ensure that we don't replace
+  // a value with something older
+
+  // Merge the data with the cache
+  // Strategy:
+  // * if nodeId is not in cache, remove the nodeId
+  // * if the nodeId is in cache, and value is None, augment cache
+  // * if the nodeid is in cache, and value is not None, pick the most recent value to augment cache
+  // Returns the values merged with cache
+  private[this] def mergeDataFromDBWithCache(fromDb: Map[NodeId, Option[NodeExpectedReports]]):IOResult[ Map[NodeId, Option[NodeExpectedReports]] ] = {
+    // In a semaphore, nothing should change the cache
+    semaphore.withPermit(
+      IOResult.effect({
+      val mergedEntries = fromDb.map {  case (nodeId, expFromDb) => cache.get(nodeId) match {
+        case None => None // nodeid not in cache, dropping
+        case Some(value) => value match { // node id in cache
+          case None => Some((nodeId -> expFromDb)) // with no value, the one from DB is better (even if none)
+          case Some(entry) => // now we must compare with the one from DB
+            expFromDb match {
+              case None => Some((nodeId -> value))
+              case Some(db) if db.beginDate.isAfter(entry.beginDate) => Some((nodeId -> expFromDb)) // data fom db is newer
```

Comment:
and here, only node with "none" in cache were called for db, so that should not happen

Reply:
query can take several seconds, cache might have been updated in between with a fresh nodeconfiguration pushed by policy generation

Reply:
I have an alternative proposal: https://github.com/ncharles/rudder/pull/10
Cache should not be updated in between, that doesn't make sense to look for empty nodeIds and then consider that that data is not the truth - or it was not usefull to check for that at first. I can't see how the DB query can't be in the semaphore if you need consistency (for ex, if you don't want to miss the fact that a cache reset was done in between).